import os
import sys
from PyQt6 import QtWidgets, uic, QtCore

# ê¸°ì¡´ main_app.pyì˜ ê²½ë¡œ ì„¤ì • ë¡œì§ ìœ ì§€
def get_resource_path(relative_path):
    if hasattr(sys, '_MEIPASS'):
        return os.path.join(sys._MEIPASS, relative_path)
    return os.path.join(os.path.abspath("."), relative_path)

root_path = os.path.dirname(os.path.abspath(__file__))
if root_path not in sys.path:
    sys.path.append(root_path)

# ì—ì´ì „íŠ¸ ë° TTS ì„í¬íŠ¸
from agents.agent_factory import create_gemini_agent
from services.tts_utils import speak

class ChatApp(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()
        
        # 1. UI ë¡œë“œ (testui.ui íŒŒì¼ ìœ„ì¹˜ í™•ì¸)
        ui_file = get_resource_path("testui.ui")
        uic.loadUi(ui_file, self)

        # 2. API í‚¤ í™•ì¸ ë° ì„¤ì • (GUI ë²„ì „)
        self.check_api_key()

        # 3. ì—ì´ì „íŠ¸ ì´ˆê¸°í™”
        try:
            self.agent = create_gemini_agent()
            self.textBrowser.append("<b>âœ¨ ì‹œìŠ¤í…œ:</b> ì œë¯¸ë‚˜ì´ê°€ ëŒ€í™” ì¤€ë¹„ë¥¼ ë§ˆì³¤ì–´!")
        except Exception as e:
            self.textBrowser.append(f"<span style='color:red;'>âŒ ì´ˆê¸°í™” ì‹¤íŒ¨: {e}</span>")

        # 4. ì‹œê·¸ë„ ì—°ê²°
        self.pushButton.clicked.connect(self.handle_send)
        self.textEdit.installEventFilter(self)
        # Enter í‚¤ ì…ë ¥ ì‹œ ì „ì†¡ (textEditê°€ ì•„ë‹ˆë¼ QLineEditì¼ ê²½ìš° ë” ì‰¬ì›€, ì—¬ê¸°ì„  ë²„íŠ¼ í´ë¦­ ìœ„ì£¼)
        
        # ğŸ’¡ ëŒ€í™” ê¸°ë¡ì„ ì €ì¥í•  ë¦¬ìŠ¤íŠ¸ ìƒì„±
        self.chat_history = []

        self.show()

    def check_api_key(self):
        """í™˜ê²½ ë³€ìˆ˜ì— í‚¤ê°€ ì—†ìœ¼ë©´ íŒì—…ì°½ìœ¼ë¡œ ì…ë ¥ë°›ìŒ"""
        if not os.getenv("GEMINI_API_KEY"):
            key, ok = QtWidgets.QInputDialog.getText(
                self, "API Key í•„ìš”", "Gemini API Keyê°€ ì—†ë„¤! ì…ë ¥í•´ì¤„ë˜?"
            )
            if ok and key:
                os.environ["GEMINI_API_KEY"] = key
            else:
                QtWidgets.QMessageBox.critical(self, "ì˜¤ë¥˜", "API í‚¤ê°€ ì—†ìœ¼ë©´ ì‹¤í–‰í•  ìˆ˜ ì—†ì–´.")
                sys.exit()

    def format_message(self, sender, message, align="left"):
        # ë°°ê²½ìƒ‰ ë° ì •ë ¬ ì„¤ì •
        if align == "right":  # ë‚˜ (ì˜¤ë¥¸ìª½)
            bg_color = "#fee500" # ë…¸ë€ìƒ‰
            table_align = "right" # ì „ì²´ í…Œì´ë¸” ë‚´ì˜ ì •ë ¬
            bubble_margin = "margin-left: 60px;" # ì™¼ìª½ ì—¬ë°±
        else:  # ì¹œêµ¬ (ì™¼ìª½)
            bg_color = "#ffffff" # í°ìƒ‰
            table_align = "left"
            bubble_margin = "margin-right: 60px;" # ì˜¤ë¥¸ìª½ ì—¬ë°±

        # 2. ì¤‘ì²© í…Œì´ë¸” êµ¬ì¡°:
        # ë°”ê¹¥ í…Œì´ë¸”(width: 100%)ì€ ì™¼ìª½/ì˜¤ë¥¸ìª½ ì •ë ¬ì„ ì¡ê³ ,
        # ì•ˆìª½ í…Œì´ë¸”(width ì—†ìŒ)ì€ ê¸€ì í¬ê¸°ë§Œí¼ë§Œ ëŠ˜ì–´ë‚©ë‹ˆë‹¤.
        return f"""
        <table width="100%" border="0" cellspacing="0" cellpadding="0" style="margin-bottom: 10px;">
            <tr>
                <td align="{table_align}">
                    <table border="0" cellspacing="0" cellpadding="0" 
                        style="background-color: {bg_color}; 
                                padding: 8px 12px;
                                border: 1px solid #d0d0d0;
                                border-radius: 12px; 
                                {bubble_margin}">
                        <tr>
                            <td style="padding: 10px; font-family: 'Malgun Gothic';">
                                <span style="font-size: 9pt; color: #555555;"><b>{sender}</b></span><br>
                                <span style="font-size: 11pt; color: #000000; line-height: 130%;">{message}</span>
                            </td>
                        </tr>
                    </table>
                </td>
            </tr>
        </table>
        """

    def handle_send(self):
        prompt = self.textEdit.toPlainText().strip()
        if not prompt: return

        # 1. ë‚´ ë©”ì‹œì§€ ì¶”ê°€ (ì˜¤ë¥¸ìª½)
        user_html = self.format_message("ë‚˜", prompt, align="right")
        self.textBrowser.append(user_html) # ğŸ’¡ ë°˜ë“œì‹œ append ì‚¬ìš©!
        self.textEdit.clear()

        # ì—ì´ì „íŠ¸ ì‹¤í–‰ ë¡œì§...
        self.process_agent_response(prompt)

    def process_agent_response(self, prompt):
        try:
            # 1. ì‚¬ìš©ì ë©”ì‹œì§€ë¥¼ ëŒ€í™” ê¸°ë¡ì— ì¶”ê°€
            self.chat_history.append({"role": "user", "content": prompt})

            # 2. ì—ì´ì „íŠ¸ í˜¸ì¶œ (ì „ì²´ ëŒ€í™” ê¸°ë¡ ì „ë‹¬)
            result = self.agent.invoke(
                {
                    "messages": self.chat_history # ğŸ’¡ ë‹¨ì¼ ë©”ì‹œì§€ê°€ ì•„ë‹Œ ì „ì²´ ê¸°ë¡ì„ ë³´ëƒ…ë‹ˆë‹¤.
                }
            )
            
            # 3. ì—ì´ì „íŠ¸ ë‹µë³€ ì¶”ì¶œ ë¡œì§ (ê¸°ì¡´ ìœ ì§€)
            raw_content = result["messages"][-1].content
            
            if isinstance(raw_content, list):
                final_answer = "".join([part.get('text', '') for part in raw_content if isinstance(part, dict)])
            else:
                final_answer = str(raw_content)

            # 4. AI ë‹µë³€ë„ ëŒ€í™” ê¸°ë¡ì— ì¶”ê°€ (ë‹¤ìŒ ëŒ€í™”ë¥¼ ìœ„í•´ ê¸°ì–µ)
            self.chat_history.append({"role": "assistant", "content": final_answer})

            # 5. UI ì—…ë°ì´íŠ¸ (ë§í’ì„  í‘œì‹œ)
            ai_html = self.format_message("ì œë¯¸ë‚˜ì´", final_answer, align="left")
            self.textBrowser.append(ai_html)
            
            # ìŠ¤í¬ë¡¤ ìµœí•˜ë‹¨ ì´ë™ ë° ìŒì„± ì¶œë ¥
            self.textBrowser.moveCursor(self.textBrowser.textCursor().MoveOperation.End)
            speak(final_answer)

        except Exception as e:
            error_msg = f"âŒ ì˜¤ë¥˜ ë°œìƒ: {e}"
            self.textBrowser.append(f"<span style='color:red;'>{error_msg}</span>")
            speak("ë¯¸ì•ˆ, ì˜¤ë¥˜ê°€ ì¢€ ë‚¬ì–´.")

if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    window = ChatApp()
    sys.exit(app.exec())